src/notebooks/controllers/controllerPreferredService.ts

```diff
@@ -397,9 +409,18 @@ export class ControllerPreferredService implements IControllerPreferredService,
                 languageInNotebookMetadata &&
                 languageInNotebookMetadata !== PYTHON_LANGUAGE &&
                 getKernelConnectionLanguage(potentialMatch) === languageInNotebookMetadata;
+            const isPythonLanguageMatch =
+                languageInNotebookMetadata &&
+                languageInNotebookMetadata === PYTHON_LANGUAGE &&
+                getKernelConnectionLanguage(potentialMatch) === languageInNotebookMetadata;
 
             // Match on our possible reasons
-            if (onlyConnection || topMatchIsPreferredInterpreter || isExactMatch || isNonPythonLanguageMatch) {
+            if (
```

Comment:
Previously if we had jsut 1 controller we recommend that as the preferred.
Even if user opens a blank notebook.

E.g. previous logic
* If you have only typescript kernel
* You open a Python notebook
* We will still recommend using typescript as the preferred
* & then when we get other controllers, we no longer update the previous controller as not preferred.

I believe that's whats going on
Difficult to repro this as its all about timing, but based on the control flow that is a definite possibility.

Desperately need unit tests for this change, will work on that

Reply:
Can we just remove the onlyConnection check totally (versus adding the isPythonLanguageMatch)? Given how push model works it seems pretty dubious in general. Too many weird cases in our code where kernels are added in out of sequence that might trigger odd suggestions here.